File: src/class072/C5_LongestNoDecreaseModifyKSubarray.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

const int MAXN = 100001;
int arr[MAXN];
int right[MAXN];
int ends[MAXN];
int n, k;

// Binary search for the longest non-increasing subsequence
int bs1(int len, int num) {
    int l = 0, r = len - 1, m, ans = -1;
    while (l <= r) {
        m = (l + r) / 2;
        if (ends[m] < num) {
            ans = m;
            r = m - 1;
        } else {
            l = m + 1;
        }
    }
    return ans;
}

// Binary search for the longest non-decreasing subsequence
int bs2(int len, int num) {
    int l = 0, r = len - 1, m, ans = -1;
    while (l <= r) {
        m = (l + r) / 2;
        if (ends[m] > num) {
            ans = m;
            r = m - 1;
        } else {
            l = m + 1;
        }
    }
    return ans;
}

// Calculates the right array which represents the length of the longest non-decreasing subsequence starting from each index
void calculateRight() {
    int len = 0;
    for (int i = n - 1; i >= 0; i--) {
        int find = bs1(len, arr[i]);
        if (find == -1) {
            ends[len++] = arr[i];
            right[i] = len;
        } else {
            ends[find] = arr[i];
            right[i] = find + 1;
        }
    }
}

int compute() {
    calculateRight();
    int len = 0;
    int ans = 0;
    for (int i = 0, j = k; j < n; i++, j++) {
        int find = bs2(len, arr[j]);
        int left = find == -1 ? len : find;
        ans = max(ans, left + k + right[j]);
        find = bs2(len, arr[i]);
        if (find == -1) {
            ends[len++] = arr[i];
        } else {
            ends[find] = arr[i];
        }
    }
    ans = max(ans, len + k);
    return ans;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    while (cin >> n >> k) {
        for (int i = 0; i < n; i++) {
            cin >> arr[i];
        }
        if (k >= n) {
            cout << n << "\n";
        } else {
            cout << compute() << "\n";
        }
    }
    return 0;
}

File: src/class072/C5_LongestNoDecreaseModifyKSubarray.py
import sys
import bisect

input = sys.stdin.read
def compute(arr, k):
    n = len(arr)
    if k >= n:
        return n
    
    # 生成辅助数组 right
    # right[j]: 一定以 arr[j] 做开头的情况下，arr[j...] 上最长不下降子序列长度是多少
    ends = []
    right = [0] * n
    for i in range(n-1, -1, -1):
        pos = bisect.bisect_left(ends, arr[i])
        if pos == len(ends):
            ends.append(arr[i])
        else:
            ends[pos] = arr[i]
        right[i] = pos + 1

    # 主逻辑部分
    ans = 0
    ends = []
    for i in range(k, n):
        if i - k >= 0:
            pos = bisect.bisect_left(ends, arr[i-k])
            if pos == len(ends):
                ends.append(arr[i-k])
            else:
                ends[pos] = arr[i-k]
        # 考虑修改 arr[i-k+1] 到 arr[i] 这个区间，与 right[i+1] 结合
        len_with_k = k + (right[i+1] if i + 1 < n else 0)
        pos = bisect.bisect_left(ends, arr[i+1] if i + 1 < n else float('inf'))
        len_without_k = pos + 1
        ans = max(ans, len_with_k, len_without_k)

    return max(ans, len(ends))

def main():
    data = input().split()
    index = 0
    while index < len(data):
        n = int(data[index])
        index += 1
        k = int(data[index])
        index += 1
        arr = list(map(int, data[index:index+n]))
        index += n
        print(compute(arr, k))

if __name__ == "__main__":
    main()
